pkg/peers/peers.go

```diff
@@ -90,33 +116,79 @@ func (p *Peers) updatePeers(ctx context.Context) {
 
 	nodes := v1.NodeList{}
 	// get some nodes, but not ourself
-	if err := p.List(readerCtx, &nodes, client.MatchingLabelsSelector{Selector: p.peerSelector}); err != nil {
+	if err := p.List(readerCtx, &nodes, client.MatchingLabelsSelector{Selector: getSelector()}); err != nil {
 		if errors.IsNotFound(err) {
 			// we are the only node at the moment... reset peerList
-			p.peersAddresses = [][]v1.NodeAddress{}
+			p.workerPeersAddresses = [][]v1.NodeAddress{}
 		}
 		p.log.Error(err, "failed to update peer list")
 		return
 	}
+
 	nodesCount := len(nodes.Items)
 	addresses := make([][]v1.NodeAddress, nodesCount)
 	for i, node := range nodes.Items {
 		addresses[i] = node.Status.Addresses
 	}
-	p.peersAddresses = addresses
+	setAddresses(addresses)
 }
 
-func (p *Peers) GetPeersAddresses() [][]v1.NodeAddress {
+func (p *Peers) GetPeersAddresses(role Role) [][]v1.NodeAddress {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
+	var addresses [][]v1.NodeAddress
+	if role == Worker {
+		addresses = p.workerPeersAddresses
+	} else {
+		addresses = p.masterPeersAddresses
+	}
 	//we don't want the caller to be able to change the addresses
 	//so we create a deep copy and return it
-	addressesCopy := make([][]v1.NodeAddress, len(p.peersAddresses))
-	for i := range p.peersAddresses {
-		addressesCopy[i] = make([]v1.NodeAddress, len(p.peersAddresses[i]))
-		copy(addressesCopy, p.peersAddresses)
+	addressesCopy := make([][]v1.NodeAddress, len(addresses))
+	for i := range addressesCopy {
+		addressesCopy[i] = make([]v1.NodeAddress, len(addresses[i]))
+		copy(addressesCopy, addresses)
 	}
 
 	return addressesCopy
 }
+
+func createSelector(hostNameToExclude string, nodeRole Role) labels.Selector {
+	var nodeTypeLabel string
+
+	switch nodeRole {
+	case Master:
+		nodeTypeLabel = GetUsedControlPlaneLabel()
+	default:
+		nodeTypeLabel = WorkerLabelName
+	}
+
+	reqNotMe, _ := labels.NewRequirement(hostnameLabelName, selection.NotEquals, []string{hostNameToExclude})
+	reqPeers, _ := labels.NewRequirement(nodeTypeLabel, selection.Exists, []string{})
+	selector := labels.NewSelector()
+	selector = selector.Add(*reqNotMe, *reqPeers)
+	return selector
+}
+
+func SetControlPlaneLabelType(node *v1.Node) {
```

Comment:
I assume this is called by 2 packages and needs sync because we don't know which one is first? I think there is a nicer way to prevent this race:

What about putting this into either another package instead, which reads the current node and role labels when the type is initialized (so no runnable for the manager, but something synchronized, so we know it runs before Start of the the others). And then peers and control plane manager get it as dependency?

Or maybe just put it (read node and labels) into New() over here?

You just need to use a reader instead of a client before Start is called, because caches are not initialized yet. But for just getting the own node this isn't an issue IMHO.

Reply:
>I assume this is called by 2 packages and needs sync because we don't know which one is first?

correct

> so no runnable for the manager, but something synchronized

In this case the k8s client will not be initialized yet so the node (which I use to decide which label to use) can't be fetched.

I think that in order to simplify the code I can modify `setNodeRole` in  `manager` not to use those methods.

Reply:
> In this case the k8s client will not be initialized yet

as metnioned in my comment already, you can use a reader instead of a client

Reply:
but the current solution looks ok to me as well